Add take_bigstring function

hcarty opened the pull request:
Useful now that the internal buffer is always a bigstring.

hcarty commented:
I think the `Bigstring.copy` usage is necessary here but I'm not 100% certain.

seliopou commented:
I've been contemplating another approach in which the underlying bigstring is exposed to a function that you pass to the parser constructor. Something like (but named differently):

```ocaml
val take_bigstring : int -> (Bigstring.t -> off:int -> len:int -> 'a) -> 'a
```

That way you don't have to perform a copy if you don't want to.  The library could also include similar parsers for `take_while`, `take_while1`, etc. What do you think of something like that?

hcarty commented:
That makes sense and gives a nice way for users to make things as efficient as they'd like.

It may be worth putting `unsafe` in the name.  If the callback does a `sub` rather than a `copy` the content could be silently overwritten.  Then again the same issue is there with any limited-lifetime resource.  `In_channel.with_file` and similar functions have the same potential issue so maybe a note in the documentation is sufficient.

I think having a copy-based `take_bigstring` (and bigstring versions of the rest of the `take_*` family) is still worth having for simplicity.  Do you agree or would you prefer to leave that decision and handling to users of the library?

hcarty commented:
Working on the `Unsafe` module - thanks for the feedback and sorry for the delay.

What do you think of keeping the easy-way-out `take_bigstring` function in addition to the more raw `Unsafe` module?

hcarty commented:
Rebased, added top-level `take_bigstring_*` functions and an `Unsafe` module.  No tests yet - I'll add those once we agree on the exposed interface and naming.

seliopou commented on lib_test/test_angstrom.ml:
```diff
@@ -3,6 +3,11 @@ open Angstrom
+  let bigstring =
+    Alcotest.testable
+      (fun fmt _bs -> Fmt.pf fmt "<bigstring>")
+      ( = )
```
This is essentially a `memcmp` on the backing buffer?

hcarty replied:
I think that's the case, though I haven't looked at the implementation.  The docs say that bigarrays support the polymorphic operators - https://caml.inria.fr/pub/docs/manual-ocaml/libref/Bigarray.html

seliopou replied:
Yup confirmed it in the runtime source code as well.

hcarty commented on lib/angstrom.mli:
```diff
@@ -376,6 +401,53 @@ val parse_string : 'a t -> string -> ('a, string) Result.result
+      containing the requested content.  [len] is the length of the requested
+      content.  [len] is guaranteed to be equal to [n]. *)
+
+  val take : int -> bigstring t
```
So something like:
```ocaml
val take_while : (char -> bool) -> (bigstring -> off:int -> len:int -> 'a) -> 'a t
```
with `off` and `len` describing the sub-bigstring which satisfies `f`?  That makes a lot of sense, and I agree regarding the safer feel.

hcarty replied:
And then `take` and `take_buffer` are equivalent - I'll keep the simpler name.

hcarty commented:
Updated according to review comments and rebased.

hcarty commented:
This version of the `Unsafe` module feels much better now - thanks for the suggestion.

hcarty commented:
@seliopou What do you think of the new `Unsafe` module implementation?

seliopou commented:
Sorry haven't had time to look but I will review this weekend!

-Spiros E.

i sent this from my Phone.

> On Oct 20, 2017, at 12:34 PM, Hezekiah M. Carty <notifications@github.com> wrote:
> 
> @seliopou What do you think of the new Unsafe module implementation?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

hcarty commented:
No problem and no rush - just checking in!

hcarty commented:
Thanks!
